src/Microsoft.Android.Build.BaseTasks/AsyncTask.cs

```diff
@@ -0,0 +1,431 @@
+// https://github.com/xamarin/Xamarin.Build.AsyncTask/blob/db4ce14dacfef47435c238b1b681c124e60ea1a0/Xamarin.Build.AsyncTask/AsyncTask.cs
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Threading;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Android.Build.Tasks
+{
+	/// <summary>
+	/// Base class for tasks that need long-running cancellable asynchronous tasks 
+	/// that don't block the UI thread in the IDE.
+	/// </summary>
+	public class AsyncTask : Task, ICancelableTask
+	{
+		readonly CancellationTokenSource cts = new CancellationTokenSource();
+		readonly Queue logMessageQueue = new Queue();
+		readonly Queue warningMessageQueue = new Queue();
+		readonly Queue errorMessageQueue = new Queue();
+		readonly Queue customMessageQueue = new Queue();
+		readonly Queue telemetryMessageQueue = new Queue ();
+		readonly ManualResetEvent logDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent errorDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent warningDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent customDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent telemetryDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent taskCancelled = new ManualResetEvent(false);
+		readonly ManualResetEvent completed = new ManualResetEvent(false);
+		bool isRunning = true;
+		object eventlock = new object();
+		int uiThreadId = 0;
+
+		/// <summary>
+		/// Indicates if the task will yield the node during tool execution.
+		/// </summary>
+		public bool YieldDuringToolExecution { get; set; }
+
+		/// <summary>
+		/// The cancellation token to notify the cancellation requests
+		/// </summary>
+		public CancellationToken CancellationToken => cts.Token;
+
+		/// <summary>
+		/// Gets the current working directory for the task, which is captured at task 
+		/// creation time from <see cref="Directory.GetCurrentDirectory"/>.
+		/// </summary>
+		protected string WorkingDirectory { get; private set; }
+
+		[Obsolete("Do not use the Log.LogXXXX from within your Async task as it will Lock the Visual Studio UI. Use the this.LogXXXX methods instead.")]
+		private new TaskLoggingHelper Log => base.Log;
+
+		/// <summary>
+		/// Initializes the task.
+		/// </summary>
+		public AsyncTask()
```

Comment:
…and otherwise reformat the code to be consistent with this repo: space before `(`, `{` on the same when appropriate (`if`, `foreach`, etc.), etc.